Sources/NetworkProtection/PacketTunnelProvider.swift

```diff
@@ -1215,6 +1219,60 @@ open class PacketTunnelProvider: NEPacketTunnelProvider {
         }
     }
 
+    lazy var failureRecoveryHandler: FailureRecoveryHandling = FailureRecoveryHandler(deviceManager: deviceManager)
+
+    private func startServerFailureMonitor() async {
+        if await serverFailureMonitor.isStarted {
+            await serverFailureMonitor.stop()
+        }
+
+        if isSubscriptionEnabled, await isEntitlementInvalid() {
+            return
+        }
+
+        await serverFailureMonitor.start { [weak self] result in
+            guard let self else {
+                return
+            }
+
+            guard case .failureDetected = result else {
+                return
+            }
+
+            Task {
+                guard let server = await self.lastSelectedServer else {
+                    os_log("🟢 No last server info found for recovery", log: .networkProtectionServerFailureRecoveryLog, type: .info)
+                    return
+                }
+                await self.startReasserting()
+                await self.serverFailureMonitor.stop()
+
+                let recoveryResult: (tunnelConfig: TunnelConfiguration, server: NetworkProtectionServer)
+                do {
+                    recoveryResult = try await self.failureRecoveryHandler.attemptRecovery(
+                        to: server,
+                        includedRoutes: self.includedRoutes ?? [],
+                        excludedRoutes: self.settings.excludedRanges,
+                        isKillSwitchEnabled: self.isKillSwitchEnabled
+                    )
+                    await self.handleFailureRecovery(tunnelConfig: recoveryResult.tunnelConfig, server: recoveryResult.server)
+                } catch {
+                    // Pixel here sent as part of https://app.asana.com/0/0/1206939413299475/f
+                    os_log("🟢 Failure recovery error: %{public}@", log: .networkProtectionServerFailureRecoveryLog, type: .error, String(reflecting: error))
+                }
+            }
```

Comment:
With the goal of trying to reduce the amount of stuff the `PacketTunnelProvider` handles, could we move this to a `ServerFailureHandler` class that's testable?

Reply:
As that’s quite a bit change and I have two PRs stacked on this one, I’d prefer to do that in another 4th PR if that’s okay. Will add you as reviewer.